File: vol1/ch7/LeetCode98.cpp
class Solution {
    TreeNode*pre;							//表示当前访问结点的前驱结点
public:
    bool isValidBST(TreeNode* root) {
        pre=NULL;								//中序首结点无前驱结点
        return inorder(root);
    }
    bool inorder(TreeNode*root) {       //判断是否为BST
        if (root==NULL)                 //空树是一棵二叉排序树
            return true;
        else {
            if(!inorder(root->left))    //判断左子树
                return false;           //若左子树不是BST时返回false
            if (pre!=NULL && pre->val>=root->val)
                return false;           //若当前结点值逆序返回false
            pre=root;                   //保存当前结点
            return inorder(root->right);   //判断右子树，返回其结果
        }
    }
};

File: vol1/ch7/LeetCode98.py
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        self.pre=None
        return self.inorder(root)

    def inorder(self,root):      #判断是否为BST
        if root==None:                 #空树是一棵二叉排序树
            return True
        else:
            if not self.inorder(root.left):    #判断左子树
                return False           #若左子树不是BST时返回false
            if self.pre!=None and self.pre.val>=root.val:
                return False           #若当前结点值逆序返回false
            self.pre=root                   #保存当前结点
            return self.inorder(root.right)   #判断右子树，返回其结果
